src/librustc_mir/interpret/eval_context.rs

```diff
@@ -116,26 +116,57 @@ pub struct LocalState<'tcx, Tag=(), Id=AllocId> {
 /// State of a local variable
 #[derive(Copy, Clone, PartialEq, Eq, Hash)]
 pub enum LocalValue<Tag=(), Id=AllocId> {
+    /// This local is not currently alive, and cannot be used at all.
     Dead,
-    // Mostly for convenience, we re-use the `Operand` type here.
-    // This is an optimization over just always having a pointer here;
-    // we can thus avoid doing an allocation when the local just stores
-    // immediate values *and* never has its address taken.
+    /// This local is alive but not yet initialized. It can be written to
+    /// but not read from or its address taken. Locals get initialized on
+    /// first write because for unsized locals, we do not know their size
+    /// before that.
+    Uninitialized,
+    /// A normal, live local.
+    /// Mostly for convenience, we re-use the `Operand` type here.
+    /// This is an optimization over just always having a pointer here;
+    /// we can thus avoid doing an allocation when the local just stores
+    /// immediate values *and* never has its address taken.
     Live(Operand<Tag, Id>),
 }
 
-impl<'tcx, Tag> LocalState<'tcx, Tag> {
+impl<Tag: Copy> LocalValue<Tag> {
+    /// The initial value of a local: ZST get "initialized" because they can be read from without
+    /// ever having been written to.
+    fn uninit_local(
+        layout: TyLayout<'_>
+    ) -> LocalValue<Tag> {
+        // FIXME: Can we avoid this ZST special case? That would likely require MIR
```

Comment:
Why do we even need to initialize it at all? ZST reads should not actually end up reading from anything

Reply:
There is code like `_0 = move _1` with both being ZST. If we never initialize `_1`, this will fail because it cannot even compute the operand (which ash to be a `MemPlace` or an `Immediate`, and an uninitialized local is neither).

Reply:
We can pull the operand from the ether. Inventing an operand if the layout is zst should be doable right where a read usually would happen.

Reply:
So, what about this?